File: 2023/day19.cpp
// https://adventofcode.com/2023/day/19

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <array>
#include <print>
#include <variant>

constexpr str start_name = "in";

struct part_t {
  std::array<i32, 4> xmas;

  fn sum() const -> i32 { return aoc::ranges::accumulate(xmas, i32{0}); }
};

struct accepted_t {};
struct rejected_t {};
using destination_t = std::variant<accepted_t, rejected_t, usize>;

constexpr let no_comparison = 4uz;

struct rule_t {
  destination_t destination;
  usize member = no_comparison;
  i32 number = 0;
};

using workflow_t = Vec<rule_t>;
using workflows_t = Vec<workflow_t>;

struct input_t {
  Vec<part_t> parts;
  workflows_t workflows;
  usize start;
};

fn rule_matches(part_t const& part, rule_t const& rule) -> bool {
  if (rule.member == no_comparison) {
    return true;
  }
  let value = part.xmas[rule.member];
  return (rule.number > 0) ? (value > rule.number) : (value < -rule.number);
}

fn check_part(part_t const& part, workflows_t const& workflows, usize start)
    -> bool {
  auto current = start;
  while (true) {
    let& workflow = workflows[current];
    for (let& rule : workflow) {
      if (!rule_matches(part, rule)) {
        continue;
      }
      if (std::holds_alternative<accepted_t>(rule.destination)) {
        return true;
      }
      if (std::holds_alternative<rejected_t>(rule.destination)) {
        return false;
      }
      current = std::get<usize>(rule.destination);
      break;
    }
  }
  return false;
}

fn solve_case1(input_t const& input) -> i32 {
  return aoc::ranges::accumulate(
      input.parts | stdv::filter([&](part_t const& part) {
        return check_part(part, input.workflows, input.start);
      }) | stdv::transform(&part_t::sum),
      i32{0});
}

fn parse(String const& filename) -> input_t {
  input_t input;

  auto name_to_id = aoc::name_to_id{};
  input.start = name_to_id.intern(start_name);

  let parse_destination = [&](str dest_str) -> destination_t {
    if (dest_str == "A") {
      return accepted_t{};
    }
    if (dest_str == "R") {
      return rejected_t{};
    }
    return name_to_id.intern(dest_str);
  };
  let parse_rule = [&](str rule_str) -> rule_t {
    // a<2006:qkq
    // rfg
    if ((rule_str.size() == 1) ||
        ((rule_str[1] != '<') && (rule_str[1] != '>'))) {
      return rule_t{.destination = parse_destination(rule_str)};
    }
    let member = [&]() -> usize {
      switch (rule_str[0]) {
        case 'x':
          return 0;
        case 'm':
          return 1;
        case 'a':
          return 2;
        case 's':
          return 3;
        default:
          AOC_UNREACHABLE("Invalid part member");
          throw 0;
      }
    }();
    let less_than = (rule_str[1] == '<');
    let[number, rule_rest_str] =
        aoc::to_number_with_rest<i32>(rule_str.substr(2));
    return rule_t{
        .destination = parse_destination(rule_rest_str.substr(1)),
        .member = member,
        .number = less_than ? -number : number,
    };
  };
  let parse_workflow = [&](str line) {
    // px{a<2006:qkq,m>2090:A,rfg}
    let[name, rules_str_raw] = aoc::split_once(line, '{');
    let rules_str = aoc::split_to_vec(
        rules_str_raw.substr(0, rules_str_raw.size() - 1), ',');
    workflow_t workflow;
    for (let rule_str : rules_str) {
      workflow.push_back(parse_rule(rule_str));
    }
    let id = name_to_id.intern(name);
    input.workflows.resize(name_to_id.new_size(input.workflows.size()));
    input.workflows[id] = std::move(workflow);
  };
  let parse_part = [&](str line) {
    // {x=787,m=2655,a=1222,s=2876}
    let[x, m, a, s] =
        aoc::split_to_array<4>(line.substr(1, line.size() - 2), ',');
    input.parts.push_back(part_t{
        .xmas =
            {
                aoc::to_number<i32>(x.substr(2)),
                aoc::to_number<i32>(m.substr(2)),
                aoc::to_number<i32>(a.substr(2)),
                aoc::to_number<i32>(s.substr(2)),
            },
    });
  };

  bool parsing_parts = false;
  for (str line : aoc::views::read_lines(filename, aoc::keep_empty{})) {
    if (line.empty()) {
      parsing_parts = true;
      continue;
    }
    if (!parsing_parts) {
      parse_workflow(line);
    } else {
      parse_part(line);
    }
  }

  return input;
}

struct range_t {
  i32 start;
  i32 end;

  fn size() const -> i64 { return static_cast<i64>(end - start + 1); }
};

fn range_product(std::array<range_t, 4> const& ranges) -> i64 {
  i64 product = 1;
  for (let& range : ranges) {
    product *= range.size();
  }
  return product;
}

fn count_accepted(workflows_t const& workflows, usize workflow_id,
                  std::array<range_t, 4> ranges) -> i64 {
  auto total = i64{};
  for (let& rule : workflows[workflow_id]) {
    auto matching = ranges;
    bool has_match = true;
    if (rule.member != no_comparison) {
      auto& match_range = matching[rule.member];
      auto& rest_range = ranges[rule.member];
      if (rule.number > 0) {
        match_range.start = std::max(match_range.start, rule.number + 1);
        rest_range.end = std::min(rest_range.end, rule.number);
      } else {
        match_range.end = std::min(match_range.end, -rule.number - 1);
        rest_range.start = std::max(rest_range.start, -rule.number);
      }
      has_match = match_range.start <= match_range.end;
    }

    if (has_match) {
      if (std::holds_alternative<accepted_t>(rule.destination)) {
        total += range_product(matching);
      } else if (std::holds_alternative<usize>(rule.destination)) {
        total += count_accepted(workflows, std::get<usize>(rule.destination),
                                matching);
      }
    }

    if ((rule.member == no_comparison) ||
        (ranges[rule.member].start > ranges[rule.member].end)) {
      break;
    }
  }
  return total;
}

fn solve_case2(input_t const& input) -> i64 {
  constexpr range_t full_range{1, 4000};
  return count_accepted(input.workflows, input.start,
                        {full_range, full_range, full_range, full_range});
}

int main() {
  std::println("Part 1");
  let example = parse("day19.example");
  AOC_EXPECT_RESULT(19114, solve_case1(example));
  let input = parse("day19.input");
  AOC_EXPECT_RESULT(509597, solve_case1(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(167409079868000, solve_case2(example));
  AOC_EXPECT_RESULT(143219569011526, solve_case2(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2023/src/day19.rs
const START_NAME: &str = "in";
const NO_COMPARISON: usize = 4;

#[derive(Clone, Copy)]
struct Part {
    xmas: [i32; 4],
}
impl Part {
    fn sum(&self) -> i32 {
        self.xmas.iter().sum()
    }
}

#[derive(Clone, Copy)]
enum Destination {
    Accepted,
    Rejected,
    Workflow(usize),
}

#[derive(Clone, Copy)]
struct Rule {
    destination: Destination,
    member: usize,
    number: i32,
}

type Workflow = Vec<Rule>;

struct Input {
    parts: Vec<Part>,
    workflows: Vec<Workflow>,
    start: usize,
}

fn parse_destination(dest_str: &str, name_to_id: &mut aoc::string::NameToId) -> Destination {
    match dest_str {
        "A" => Destination::Accepted,
        "R" => Destination::Rejected,
        _ => Destination::Workflow(name_to_id.intern(dest_str)),
    }
}

fn parse_rule(rule_str: &str, name_to_id: &mut aoc::string::NameToId) -> Rule {
    // a<2006:qkq
    // rfg
    let bytes = rule_str.as_bytes();
    if rule_str.len() == 1 || (bytes[1] != b'<' && bytes[1] != b'>') {
        return Rule {
            destination: parse_destination(rule_str, name_to_id),
            member: NO_COMPARISON,
            number: 0,
        };
    }
    let member = match bytes[0] {
        b'x' => 0,
        b'm' => 1,
        b'a' => 2,
        b's' => 3,
        _ => unreachable!("Invalid part member"),
    };
    let less_than = bytes[1] == b'<';
    let (number_str, dest_str) = rule_str[2..].split_once(':').unwrap();
    let number: i32 = number_str.parse().unwrap();
    Rule {
        destination: parse_destination(dest_str, name_to_id),
        member,
        number: if less_than { -number } else { number },
    }
}

fn parse(filename: &str) -> Input {
    let lines = aoc::file::read_lines(filename);

    let mut name_to_id = aoc::string::NameToId::new();
    let start = name_to_id.intern(START_NAME);

    let mut workflows: Vec<Workflow> = Vec::new();
    let mut parts: Vec<Part> = Vec::new();
    let mut parsing_parts = false;

    for line in &lines {
        if line.is_empty() {
            parsing_parts = true;
            continue;
        }
        if !parsing_parts {
            // px{a<2006:qkq,m>2090:A,rfg}
            let (name, rules_str) = line.split_once('{').unwrap();
            let rules_str = &rules_str[..rules_str.len() - 1];
            let workflow: Workflow = rules_str
                .split(',')
                .map(|rule_str| parse_rule(rule_str, &mut name_to_id))
                .collect();
            let id = name_to_id.intern(name);
            workflows.resize(name_to_id.new_len(workflows.len()), Vec::new());
            workflows[id] = workflow;
        } else {
            // {x=787,m=2655,a=1222,s=2876}
            let nums: Vec<i32> = line[1..line.len() - 1]
                .split(',')
                .map(|n| n[2..].parse().unwrap())
                .collect();
            parts.push(Part {
                xmas: [nums[0], nums[1], nums[2], nums[3]],
            });
        }
    }

    Input {
        parts,
        workflows,
        start,
    }
}

fn rule_matches(part: &Part, rule: &Rule) -> bool {
    if rule.member == NO_COMPARISON {
        return true;
    }
    let value = part.xmas[rule.member];
    if rule.number > 0 {
        value > rule.number
    } else {
        value < -rule.number
    }
}

fn check_part(part: &Part, workflows: &[Workflow], start: usize) -> bool {
    let mut current = start;
    loop {
        for rule in &workflows[current] {
            if !rule_matches(part, rule) {
                continue;
            }
            match rule.destination {
                Destination::Accepted => return true,
                Destination::Rejected => return false,
                Destination::Workflow(id) => {
                    current = id;
                    break;
                }
            }
        }
    }
}

fn solve_case1(input: &Input) -> i32 {
    input
        .parts
        .iter()
        .filter(|part| check_part(part, &input.workflows, input.start))
        .map(Part::sum)
        .sum()
}

#[derive(Clone, Copy)]
struct Range {
    start: i32,
    end: i32,
}

impl Range {
    fn size(self) -> i64 {
        (self.end - self.start + 1) as i64
    }
}

fn range_product(ranges: &[Range; 4]) -> i64 {
    ranges.iter().map(|range| range.size()).product()
}

fn count_accepted(workflows: &[Workflow], workflow_id: usize, mut ranges: [Range; 4]) -> i64 {
    let mut total = 0i64;
    for rule in &workflows[workflow_id] {
        let mut matching = ranges;
        let mut has_match = true;
        if rule.member != NO_COMPARISON {
            if rule.number > 0 {
                matching[rule.member].start = matching[rule.member].start.max(rule.number + 1);
                ranges[rule.member].end = ranges[rule.member].end.min(rule.number);
            } else {
                matching[rule.member].end = matching[rule.member].end.min(-rule.number - 1);
                ranges[rule.member].start = ranges[rule.member].start.max(-rule.number);
            }
            has_match = matching[rule.member].start <= matching[rule.member].end;
        }

        if has_match {
            match rule.destination {
                Destination::Accepted => total += range_product(&matching),
                Destination::Rejected => {}
                Destination::Workflow(id) => total += count_accepted(workflows, id, matching),
            }
        }

        if rule.member == NO_COMPARISON || ranges[rule.member].start > ranges[rule.member].end {
            break;
        }
    }
    total
}

fn solve_case2(input: &Input) -> i64 {
    let full_range = Range {
        start: 1,
        end: 4000,
    };
    count_accepted(&input.workflows, input.start, [full_range; 4])
}

fn main() {
    println!("Part 1");
    let example = parse("day19.example");
    aoc::expect_result!(19114, solve_case1(&example));
    let input = parse("day19.input");
    aoc::expect_result!(509597, solve_case1(&input));

    println!("Part 2");
    aoc::expect_result!(167409079868000, solve_case2(&example));
    aoc::expect_result!(143219569011526, solve_case2(&input));
}
